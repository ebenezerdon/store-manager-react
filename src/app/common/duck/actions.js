import types from './types';

const setFetchCurrentUserState = fetchCurrentUserState => ({
  type: types.SET_FETCH_CURRENT_USER_STATE,
  fetchCurrentUserState,
});

const setFetchCurrentUserError = errorMessage => ({
  type: types.SET_FETCH_CURRENT_USER_ERROR,
  errorMessage,
});

const setFetchAllUserState = fetchAllUserState => ({
  type: types.SET_FETCH_ALL_USER_STATE,
  fetchAllUserState,
});

const setFetchAllUserError = errorMessage => ({
  type: types.SET_FETCH_ALL_USER_ERROR,
  errorMessage,
});

const setFetchSaleRecordState = fetchSaleRecordState => ({
  type: types.SET_FETCH_SALE_RECORD_STATE,
  fetchSaleRecordState,
});

const setFetchSaleRecordError = errorMessage => ({
  type: types.SET_FETCH_SALE_RECORD_ERROR,
  errorMessage,
});

const setFetchProductsState = fetchProductsState => ({
  type: types.SET_FETCH_PRODUCTS_STATE,
  fetchProductsState,
});

const setFetchProductsError = errorMessage => ({
  type: types.SET_FETCH_PRODUCTS_ERROR,
  errorMessage,
});

const addUserData = userData => ({
  type: types.ADD_USER_DATA,
  userData,
});

const addSaleRecord = saleRecord => ({
  type: types.ADD_SALE_RECORD,
  saleRecord,
});

const addProducts = products => ({
  type: types.ADD_PRODUCTS,
  products,
});

export default {
  setFetchCurrentUserState,
  setFetchCurrentUserError,
  setFetchAllUserState,
  setFetchAllUserError,
  setFetchSaleRecordState,
  setFetchSaleRecordError,
  setFetchProductsState,
  setFetchProductsError,
  addUserData,
  addSaleRecord,
  addProducts
};
