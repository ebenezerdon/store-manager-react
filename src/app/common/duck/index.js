export {
  doFetchCurrentUser,
  doFetchAllUsers,
  doFetchSaleRecord,
  doFetchProducts
} from './operations';
export { default as fetchAllDataReducer } from './reducers';
export { default as constants } from './constants';
export { default as types } from './types';
export { default as actions } from './actions';
