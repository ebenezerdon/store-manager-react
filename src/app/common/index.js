export { fetchAllDataReducer } from './duck';
export { constants } from './duck';
export { EllipsisLoaderComponent } from './loaders';

export {
  doFetchCurrentUser,
  doFetchAllUsers,
  doFetchSaleRecord,
  doFetchProducts
} from './duck';
