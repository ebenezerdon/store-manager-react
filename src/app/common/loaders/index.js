export { default as EllipsisLoaderComponent } from './EllipsisLoaderComponent';
