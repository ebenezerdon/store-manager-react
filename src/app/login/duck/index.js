export { default as loginReducer } from './reducers';
export { default as constants } from './constants';
export { default as doLogin } from './operations';
export { default as actions } from './actions';
export { default as types } from './types';
