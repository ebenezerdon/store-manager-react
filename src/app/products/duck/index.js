export { addProductReducer, checkoutReducer } from './reducers';
export { default as constants } from './constants';
export { doAddProduct, doAddToCart, doCheckout } from './operations';
export { default as actions } from './actions';
export { default as types } from './types';
