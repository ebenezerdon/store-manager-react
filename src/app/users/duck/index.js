export { default as addUserReducer } from './reducers';
export { default as constants } from './constants';
export { default as doAddUser } from './operations';
export { default as actions } from './actions';
export { default as types } from './types';
