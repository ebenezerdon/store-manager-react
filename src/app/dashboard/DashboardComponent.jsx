import React, { Fragment } from 'react';
import lifecycle from 'react-pure-lifecycle';
import { Link } from 'react-router-dom';
import { constants } from '../common';
import { EllipsisLoaderComponent } from '../common'
import '../style/admindashboard.css';

const methods = {
  componentWillMount({
    fetchCurrentUser,
    fetchallUsers,
    fetchSaleRecord,
    fetchProducts,
    userData,
    saleRecord,
    products
  }) {
    if (!userData.currentUser) {
      fetchCurrentUser();
      fetchallUsers();
      fetchSaleRecord();
      fetchProducts();
    }
  }
};

const DashboardComponent = ({
  userData,
  saleRecord,
  products,
  fetchProductsState,
  fetchSaleRecordState
}) => {
  const userProfile = () => {
    const { currentUser } = userData;
    return (
      <figure className="profile">
        <div className="profile-image">
          <img src= {currentUser.userimage} alt="User Image" />
        </div>
        <figcaption>
          <h3>{currentUser.fullname}</h3>
          <h4>Store {currentUser.role}</h4>
        </figcaption>
      </figure>
    )
  }

  const latestProducts = () => {
    const { data } = products;
    const sortedData = [...data].reverse();
    return (
      sortedData.slice(0, 4).map(product => (
        <Link className="col-md-6 col-sm-12"
          to={`/product/${product.id}`}>
          <img src={product.productimage} width='300' />
        </Link>
      ))
    )
  }

  const storeAttendants = () => {
    const { allUsers } = userData;
    return (
      allUsers.slice(0, 4).map(user => (
        <li>
          <Link className="list" to={`/profile/${user.id}`}>{user.fullname}</Link>
        </li>
      ))
    )
  }

  const sales = () => {
    const { data } = saleRecord;
    return (
      data.slice(0, 4).map(saleRecord => (
        <li>
          <Link className="list" to={`/product/${saleRecord.productid}`}>{saleRecord.productname}</Link>
        </li>
      ))
    )
  }

  return(
    <>
      <div id="centerLoader">
        {fetchProductsState === constants.FETCHING_PRODUCTS &&
          <EllipsisLoaderComponent/>
        }
      </div>
      <div className="row" id="dashboardColumns">
        <div className='product'>
          {userData.currentUser && userProfile()}
        </div>
        <div className='product'>
          <div className="product-item tp">
            <h3>Latest Products</h3>
            <div className="t-products" id="latestProducts">
              {products.data && latestProducts()}
            </div>
            <Link className="green-button" to="/products">See All Products</Link>
          </div>
        </div>

        <div className='product'>
          <div className="product-item tp">
            <h3>Store Attendants</h3>
            <ul className="user-list">
              {userData.allUsers && storeAttendants()}
            </ul>
            <Link className="green-button" to="/users">See All</Link>
          </div>
        </div>

        <div className='product'>
          <div className="product-item tp">
            <h3>New Sales</h3>
            <ul className="user-list">
              {saleRecord.data && sales()}
            </ul>
            <Link className="green-button" to="/sales">See All Sale Record</Link>
          </div>
        </div>
      </div>
    </>
  )
}

export default lifecycle(methods)(DashboardComponent);
