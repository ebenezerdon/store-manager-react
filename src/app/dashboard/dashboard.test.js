import React from 'react';
import thunk from 'redux-thunk';
import axios from 'axios';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import DashboardComponent from './DashboardComponent';
import { actions, constants, fetchAllDataReducer, types } from '../common/duck';
import { EllipsisLoaderComponent } from '../common/loaders';
import {
  DashboardContainer,
  mapDispatchToProps,
  mapStateToProps,
} from './DashboardContainer';

jest.mock('axios');

const {
  setFetchCurrentUserState,
  setFetchCurrentUserError,
  setFetchAllUserState,
  setFetchAllUserError,
  setFetchSaleRecordState,
  setFetchSaleRecordError,
  setFetchProductsState,
  setFetchProductsError,
  addUserData,
  addSaleRecord,
  addProducts
} = actions;

describe('ListArticle Component Action', () => {
  it('it should setFetchCurrentUserState', () => {
    const action = setFetchCurrentUserState(constants.FETCHING_CURRENT_USER);
    expect(action).toEqual({
      type: types.SET_FETCH_CURRENT_USER_STATE,
      fetchCurrentUserState: constants.FETCHING_CURRENT_USER,
    });
  });

  it('it should setFetchCurrentUserError', () => {
    const action = setFetchCurrentUserError(constants.FETCH_CURRENT_USER_ERROR);
    expect(action).toEqual({
      type: types.SET_FETCH_CURRENT_USER_ERROR,
      errorMessage: constants.FETCH_CURRENT_USER_ERROR,
    });
  });
  it('it should setFetchAllUserState', () => {
    const action = setFetchAllUserState(constants.FETCHING_ALL_USERS);
    expect(action).toEqual({
      type: types.SET_FETCH_ALL_USER_STATE,
      fetchAllUserState: constants.FETCHING_ALL_USERS,
    });
  });

  it('it should setFetchAllUserError', () => {
    const action = setFetchAllUserError(constants.FETCH_ALL_USER_ERROR);
    expect(action).toEqual({
      type: types.SET_FETCH_ALL_USER_ERROR,
      errorMessage: constants.FETCH_ALL_USER_ERROR,
    });
  });
  it('it should setFetchSaleRecordState', () => {
    const action = setFetchSaleRecordState(constants.FETCHING_SALE_RECORD);
    expect(action).toEqual({
      type: types.SET_FETCH_SALE_RECORD_STATE,
      fetchSaleRecordState: constants.FETCHING_SALE_RECORD,
    });
  });

  it('it should setFetchSaleRecordError', () => {
    const action = setFetchSaleRecordError(constants.FETCH_SALE_RECORD_ERROR);
    expect(action).toEqual({
      type: types.SET_FETCH_SALE_RECORD_ERROR,
      errorMessage: constants.FETCH_SALE_RECORD_ERROR,
    });
  });
  it('it should setFetchProductsState', () => {
    const action = setFetchProductsState(constants.FETCH_PRODUCTS_SUCCESS);
    expect(action).toEqual({
      type: types.SET_FETCH_PRODUCTS_STATE,
      fetchProductsState: constants.FETCH_PRODUCTS_SUCCESS,
    });
  });

  it('it should setFetchProductsError', () => {
    const action = setFetchProductsError(constants.FETCH_PRODUCTS_ERROR);
    expect(action).toEqual({
      type: types.SET_FETCH_PRODUCTS_ERROR,
      errorMessage: constants.FETCH_PRODUCTS_ERROR,
    });
  });

  it('it should addUserData', () => {
    const action = addUserData('Data');
    expect(action).toEqual({
      type: types.ADD_USER_DATA,
      userData: 'Data',
    });
  });

  it('it should addSaleRecord', () => {
    const action = addSaleRecord('Data');
    expect(action).toEqual({
      type: types.ADD_SALE_RECORD,
      saleRecord: 'Data',
    });
  });

  it('it should addProducts', () => {
    const action = addProducts('Data');
    expect(action).toEqual({
      type: types.ADD_PRODUCTS,
      products: 'Data',
    });
  });
});

describe('DashboardContainer', () => {
  it('should show initial state', () => {
    const initialState = {
      fetchAllData: {
        fetchCurrentUserState: '',
        fetchAllUserState: '',
        fetchSaleRecordState: '',
        fetchProductsState: '',
        userData: '',
        saleRecord: '',
        products: '',
        errorMessage: ''
      },
      location: { pathname: '' },
    };
    expect(mapStateToProps(initialState).fetchAllUserState).toEqual('');
    expect(mapStateToProps(initialState).errorMessage).toEqual('');
  });

  it('should dispatch action', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).fetchCurrentUser();
    mapDispatchToProps(dispatch).fetchallUsers();
    mapDispatchToProps(dispatch).fetchSaleRecord();
    mapDispatchToProps(dispatch).fetchProducts();
    expect(typeof dispatch.mock.calls[0][0]).toEqual('function');
  });
});

describe('fetchAllDataReducer', () => {
  it('should setup default state values', () => {
    const state = fetchAllDataReducer(undefined, {
      type: '@@INIT',
    });
    expect(state.fetchCurrentUserState).toEqual('');
  });

  it('it should SET_FETCH_CURRENT_USER_STATE', () => {
    const action = {
      type: types.SET_FETCH_CURRENT_USER_STATE,
      fetchCurrentUserState: constants.FETCHING_CURRENT_USER,
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.fetchCurrentUserState).toEqual(constants.FETCHING_CURRENT_USER);
  });

  it('it should change the fetch CURRENT_USER error message', () => {
    const action = {
      type: types.SET_FETCH_CURRENT_USER_ERROR,
      errorMessage: 'failed to fetch',
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.errorMessage).toEqual(action.errorMessage);
  });
  it('it should change the FETCH_ALL_USER_STATE', () => {
    const action = {
      type: types.SET_FETCH_ALL_USER_STATE,
      fetchAllUserState: constants.FETCHING_ALL_USERS,
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.fetchAllUserState).toEqual(constants.FETCHING_ALL_USERS);
  });

  it('it should change the fetch ALL_USERS error message', () => {
    const action = {
      type: types.SET_FETCH_ALL_USER_ERROR,
      errorMessage: 'failed to fetch',
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.errorMessage).toEqual(action.errorMessage);
  });
  it('it should SET_FETCH_SALE_RECORD_STATE', () => {
    const action = {
      type: types.SET_FETCH_SALE_RECORD_STATE,
      fetchSaleRecordState: constants.FETCHING_SALE_RECORD,
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.fetchSaleRecordState).toEqual(constants.FETCHING_SALE_RECORD);
  });

  it('it should change the FETCH_SALE_RECORD error message', () => {
    const action = {
      type: types.SET_FETCH_SALE_RECORD_ERROR,
      errorMessage: 'failed to fetch',
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.errorMessage).toEqual(action.errorMessage);
  });
  it('it should SET_FETCH_PRODUCTS_STATE', () => {
    const action = {
      type: types.SET_FETCH_PRODUCTS_STATE,
      fetchProductsState: constants.FETCHING_PRODUCTS,
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.fetchProductsState).toEqual(constants.FETCHING_PRODUCTS);
  });

  it('it should change the fetch PRODUCTS error message', () => {
    const action = {
      type: types.SET_FETCH_PRODUCTS_ERROR,
      errorMessage: 'failed to fetch',
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.errorMessage).toEqual(action.errorMessage);
  });

  it('it should ADD_USER_DATA', () => {
    const action = {
      type: types.ADD_USER_DATA,
      userData: {Data: ['Data']},
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.userData).toEqual(action.userData);
  });

  it('it should ADD_SALE_RECORD', () => {
    const action = {
      type: types.ADD_SALE_RECORD,
      saleRecord: {Data: ['Data']},
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.saleRecord).toEqual(action.saleRecord);
  });

  it('it should ADD_PRODUCTS', () => {
    const action = {
      type: types.ADD_PRODUCTS,
      products: {Data: ['Data']},
    };
    const state = fetchAllDataReducer(undefined, action);
    expect(state.products).toEqual(action.products);
  });

  it('should render the component', () => {
    const fetchCurrentUser = jest.fn();
    const fetchallUsers = jest.fn();
    const fetchSaleRecord = jest.fn();
    const fetchProducts = jest.fn();
    const userData = {
      currentUser: [{
        userimage: 'userimage',
        fullname: 'fullname',
        role: 'role'
      }],
    };
    const saleRecord = {
      data: [{sale: 'sale'}],
    };
    const products = {
      data: [{product: 'product'}],
    };
    const component = shallow(
      <DashboardComponent
        {...{
          fetchCurrentUser,
          fetchallUsers,
          fetchSaleRecord,
          fetchProducts,
          userData,
          saleRecord,
          products
        }}
      />,
    );
    expect(component.exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });
});

describe('Connected DashboardComponent Component Dispatches Success', () => {
  let wrapper;
  const initialState = {
    fetchAllData: {
      fetchCurrentUserState: '',
      fetchAllUserState: '',
      fetchSaleRecordState: '',
      fetchProductsState: '',
      userData: '',
      saleRecord: '',
      products: '',
      errorMessage: ''
    },
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore(initialState);
  beforeEach(() => {
    const response = {
      data: [],
    };
    axios.get.mockImplementation(() =>
      Promise.resolve({ data: { ...response } }),
    );
    const userData = [];
    const saleRecord = [];
    const products = [];
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <DashboardContainer
            userData={userData}
            saleRecord={saleRecord}
            products = {products} />
        </MemoryRouter>
      </Provider>,
    );
  });

  it('it should render the connected component', () => {
    expect(wrapper.find(DashboardComponent).length).toEqual(1);
  });

  it('it should dispatch fetchCurrentUser action', () => {
    const storeState = store.getState();
    const storeActions = store.getActions();
    expect(storeActions[0].fetchCurrentUserState).toEqual('FETCHING_CURRENT_USER');
    expect(storeActions[1].fetchAllUserState).toEqual('FETCHING_ALL_USERS');
    expect(storeActions[2].fetchSaleRecordState).toEqual('FETCHING_SALE_RECORD');
    expect(storeActions[3].fetchProductsState).toEqual('FETCHING_PRODUCTS');
    expect(storeActions[4].type).toEqual('ADD_USER_DATA');
  });
});
