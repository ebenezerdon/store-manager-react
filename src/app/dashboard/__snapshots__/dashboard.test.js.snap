// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAllDataReducer should render the component 1`] = `
<DashboardComponent
  fetchCurrentUser={[MockFunction]}
  fetchProducts={[MockFunction]}
  fetchSaleRecord={[MockFunction]}
  fetchallUsers={[MockFunction]}
  products={
    Object {
      "data": Array [
        Object {
          "product": "product",
        },
      ],
    }
  }
  saleRecord={
    Object {
      "data": Array [
        Object {
          "sale": "sale",
        },
      ],
    }
  }
  userData={
    Object {
      "currentUser": Array [
        Object {
          "fullname": "fullname",
          "role": "role",
          "userimage": "userimage",
        },
      ],
    }
  }
/>
`;
